[clusteragent/admission/webhook] Add selectors needed for AKS

davidor opened the pull request:
### What does this PR do?

This PR fixes a problem that happens when enabling the DCA admission controller on Azure AKS.

AKS automatically updates our webhooks adding a selector that excludes the "control-plane" namespace. This creates conflicts when we reconcile the webhooks in the controller and spams the logs with messages like:

`Couldn't reconcile Webhook datadog-webhook: Operation cannot be
fulfilled on mutatingwebhookconfigurations.admissionregistration.k8s.io
"datadog-webhook": the object has been modified; please apply your
changes to the latest version and try again`.

References:
- https://docs.microsoft.com/en-us/azure/aks/faq#can-i-use-admission-controller-webhooks-on-aks
- https://github.com/Azure/AKS/issues/1771

### Describe how to test/QA your changes

Deploy on AKS with the DCA admission controller enabled. If you're using the Helm charts:
```
...
clusterAgent:
  replicas: 2
  admissionController:
    enabled: true
    mutateUnlabelled: true
...
```

Also, set the new config option `DD_ADMISSION_CONTROLLER_ADD_AKS_SELECTORS` to true.

Check that there are not messages like the one posted above being logged continuously. Also check that the admission controller is still working as expected.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

davidor committed:
[clusteragent/admission/webhook] Add selectors needed for AKS

This fixes a problem that happens when enabling the DCA admission
controller on Azure AKS.

AKS automatically updates our webhooks adding a selector that excludes
the "control-plane" namespace. This creates conflicts when we reconcile
the webhooks in the controller and spams the logs with messages like:

`Couldn't reconcile Webhook datadog-webhook: Operation cannot be
fulfilled on mutatingwebhookconfigurations.admissionregistration.k8s.io
"datadog-webhook": the object has been modified; please apply your
changes to the latest version and try again`.

References:
- https://docs.microsoft.com/en-us/azure/aks/faq#can-i-use-admission-controller-webhooks-on-aks
- https://github.com/Azure/AKS/issues/1771

davidor committed:
Add release note

davidor commented:
TODO: We should add a feature flag so this can be enabled optionally or investigate if there's a reliable way of knowing if the DCA is deployed on AKS.

ahmed-mez commented on pkg/clusteragent/admission/controllers/webhook/common.go:
```diff
@@ -28,12 +30,32 @@ func buildLabelSelector() *metav1.LabelSelector {
-		MatchLabels: map[string]string{
-			common.EnabledLabelKey: "true",
-		},
+	labelSelector.MatchExpressions = append(
+		labelSelector.MatchExpressions,
+		azureAKSLabelSelectorRequirement(),
+	)
```
Based on [azure's docs](https://docs.microsoft.com/en-us/azure/aks/faq#can-i-use-admission-controller-webhooks-on-aks), this should only be added as a namespace selector, while this function `buildLabelSelector` return value can be used as object selector (default), so we're adding the rule unnecessarily for object selectors.
I think we should adapt the code to only add this aks rule to the namespace selector.

davidor replied:
Yes. You're right :+1:

davidor committed:
[config] Add kubernetes_distribution option

davidor committed:
review: don't add constraint to object selector

davidor committed:
review: case-insensitive config option

davidor commented on pkg/clusteragent/admission/controllers/webhook/common.go:
```diff
@@ -28,12 +30,59 @@ func buildLabelSelector() *metav1.LabelSelector {
 }
+
+func runningOnAKS() bool {
+	return config.Datadog.GetString("kubernetes_distribution") == "aks"
```
Fixed: https://github.com/DataDog/datadog-agent/pull/11367/commits/717fc363564cb51d9dab6d6b7d5cec5c942b42a9

djmitche commented on pkg/config/config_template.yaml:
```diff
@@ -2520,6 +2520,15 @@ api_key:
+## @env DD_KUBERNETES_DISTRIBUTION - string - optional - default: ""
+# The Kubernetes distribution where the agent is deployed. This automatically
+# applies options needed to make the agent run correctly on each distribution.
+# For now, this is only needed when running on AKS. When that's the case, set
+# this option to "aks".
+#
```
What would it be set to in other circumstances?  It seems like this has been added as a way to say "should we apply the AKS-specific workaround" -- maybe a more specific name would be better?

davidor replied:
Right. I'll change this. I got similar feedback from @vboulineau

davidor replied:
I added a new commit to change this: https://github.com/DataDog/datadog-agent/pull/11367/commits/a51cb37631c82b3d4473e0cca1feb55275705f0b

davidor committed:
review: use specific config option

davidor commented:
I added a new commit (https://github.com/DataDog/datadog-agent/pull/11367/commits/a51cb37631c82b3d4473e0cca1feb55275705f0b) to change the config option to a more specific one as suggested in https://github.com/DataDog/datadog-agent/pull/11367#discussion_r833516589 

I also rebased on top of `main` to fix conflicts.

davidor committed:
review: move release note to DCA dir

djmitche reviewed (approved):
Changes look good

davidor commented:
Rebased on top of `main` again.
